File: SPP-WEB/app/Http/Controllers/PembayaranController.php
<?php

namespace App\Http\Controllers;

use App\Models\Kelas;
use App\Models\Pembayaran;
use App\Models\Siswa;
use Barryvdh\DomPDF\Facade\Pdf;
use Illuminate\Http\Request;
use Illuminate\Support\Facades\Auth;
use Illuminate\Support\Facades\DB;

class PembayaranController extends Controller
{
    public function index()
    {
        $siswa = Siswa::with('kelas')->with('spp')->get();
        $pembayaran = Pembayaran::all();
        $kelas = Kelas::all();
        $bulan = ['Juli', 'Agustus', 'September', 'Oktober', 'November', 'Desember', 'Januari', 'Februari', 'Maret', 'April', 'Mei', 'Juni'];

        return view('petugas.pembayaran.index', compact('siswa', 'bulan', 'pembayaran', 'kelas'));
    }

    public function cari(Request $request)
    {
        $keyword = $request->keyword;
        $id_kelas = $request->id_kelas;
        $siswa = Siswa::with('kelas')->with('spp')->where('nama', 'LIKE', '%' . $keyword . '%')->get();
        if ($id_kelas != "semua") {
            $siswa = $siswa->where('id_kelas', $id_kelas);
        }
        $pembayaran = Pembayaran::all();
        $kelas = Kelas::all();
        $bulan = ['Juli', 'Agustus', 'September', 'Oktober', 'November', 'Desember', 'Januari', 'Februari', 'Maret', 'April', 'Mei', 'Juni'];

        return view('petugas.pembayaran.index', compact('siswa', 'keyword', 'bulan', 'pembayaran', 'kelas', 'id_kelas'));
    }

    public function bayar(Request $request, $nisn)
    {
        $request->validate([
            'bulan_dibayar' => 'required|array',
            'bulan_dibayar.*' => 'string'
        ]);

        $siswa = Siswa::find($nisn);
        if (!$siswa) {
            return redirect()->back()->with('error', 'Data siswa tidak ditemukan.');
        }

        $id_spp = $siswa->id_spp;
        $nominal_spp = $siswa->spp->nominal;

        foreach ($request->bulan_dibayar as $bulan) {
            $cek = Pembayaran::where('nisn', $siswa->nisn)
                ->where('bulan_dibayar', $bulan)
                ->where('tahun_dibayar', $siswa->spp->tahun)
                ->exists();

            if ($cek) {
                continue;
            }

            Pembayaran::create([
                'id_petugas' => Auth::guard('petugas')->user()->id_petugas,
                'nisn' => $siswa->nisn,
                'tgl_bayar' => now(),
                'bulan_dibayar' => $bulan,
                'tahun_dibayar' => $siswa->spp->tahun,
                'id_spp' => $id_spp,
                'jumlah_bayar' => $nominal_spp,
            ]);
        }

        return redirect('/pembayaran/detail/' . $nisn)->with('success', 'Pembayaran berhasil ditambahkan!');
    }

    public function riwayat()
    {
        $pembayaran = collect(DB::select('SELECT * FROM riwayat_pembayaran'));
        $kelas = Kelas::all();

        return view('petugas.pembayaran.riwayat', compact('pembayaran', 'kelas'));
    }

    public function cariRiwayat(Request $request)
    {
        $keyword = $request->keyword;
        $id_kelas = $request->id_kelas;
        $kelas = Kelas::all();
        $pembayaran = collect(DB::select("SELECT * FROM riwayat_pembayaran WHERE nama LIKE '%" . $keyword . "%'"));
        if ($id_kelas != "semua") {
            $pembayaran = collect(DB::select("SELECT * FROM riwayat_pembayaran WHERE nama LIKE '%" . $keyword . "%' AND id_kelas = '" . $id_kelas . "'"));
        }

        return view('petugas.pembayaran.riwayat', compact('pembayaran', 'keyword', 'id_kelas', 'kelas'));
    }

    public function detail($nisn)
    {
        $s = Siswa::with('kelas')->with('spp')->where('nisn', $nisn)->first();
        $pembayaran = Pembayaran::with('petugas')->where('nisn', $nisn)->get();
        $siswa = collect(DB::select("SELECT * FROM riwayat_pembayaran WHERE nisn = '" . $nisn . "'"))->first();
        $bulan = ['Juli', 'Agustus', 'September', 'Oktober', 'November', 'Desember', 'Januari', 'Februari', 'Maret', 'April', 'Mei', 'Juni'];

        return view('petugas.pembayaran.detail', compact('pembayaran', 'siswa', 'bulan', 's'));
    }

    public function cetak($id)
    {
        $pembayaran = Pembayaran::with('siswa')->with('spp')->with('petugas')->find($id);

        $pdf = Pdf::loadView('petugas.pembayaran.cetak', compact('pembayaran'))->setPaper('a5', 'portrait');
        return $pdf->stream('invoice.pdf');
    }

    public function cetakKartu($nisn)
    {
        $pembayaran = Pembayaran::with('petugas')->where('nisn', $nisn)->get();
        $siswa = collect(DB::select("SELECT * FROM riwayat_pembayaran WHERE nisn = '" . $nisn . "'"))->first();
        $bulan = ['Juli', 'Agustus', 'September', 'Oktober', 'November', 'Desember', 'Januari', 'Februari', 'Maret', 'April', 'Mei', 'Juni'];

        return view('petugas.pembayaran.kartu', compact('pembayaran', 'siswa', 'bulan'));
    }

    public function hapus($id)
    {
        Pembayaran::find($id)->delete();
        return redirect()->back()->with('success', 'Hapus Transaksi Berhasil!');
    }

    public function gagal()
    {
        $kelas = Kelas::all();
        $pembayaran = Pembayaran::onlyTrashed()->with('siswa')->get();

        return view('petugas.pembayaran.dihapus', compact('kelas', 'pembayaran'));
    }

    public function filterGagal(Request $request)
    {
        $dari = $request->dari;
        $sampai = $request->sampai;

        if (!$dari && !$sampai) {
            return redirect('/pembayaran/riwayat/gagal');
        }

        $pembayaran = Pembayaran::onlyTrashed()
            ->with('siswa')
            ->whereBetween('tgl_bayar', [$dari, $sampai])
            ->get();

        return view('petugas.pembayaran.dihapus', compact('pembayaran', 'dari', 'sampai'));
    }

}

File: SPP-Android/backend/API-SPP/API-SPP/Controllers/PembayaranController.cs
using API_SPP.Helpers;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Newtonsoft.Json;
using System.Collections.Generic;
using System.Data;

namespace API_SPP.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    //[Authorize]
    public class PembayaranController : ControllerBase
    {
        [HttpGet]
        public IActionResult GetPembayaran([FromQuery] string nama, [FromQuery] string kelas)
        {
            DB db = new DB();

            string query = "SELECT * FROM list_pembayaran WHERE 1=1";

            if (!string.IsNullOrEmpty(nama))
                query += $" AND nama LIKE '%{nama}%'";

            if (!string.IsNullOrEmpty(kelas))
                query += $" AND id_kelas = '{kelas}'";

            DataTable siswaTable = db.Query(query);
            string siswaList = JsonConvert.SerializeObject(siswaTable);

            DataTable kelasTable = db.Query("SELECT * FROM kelas");
            string kelasList = JsonConvert.SerializeObject(kelasTable);

            return Ok(new
            {
                siswa = siswaList,
                kelas = kelasList
            });
        }

        [HttpGet("detail/{nisn}")]
        public IActionResult Detail(string nisn)
        {
            DB db = new DB();

            DataTable dtPembayaran = db.Query(
                $"SELECT * FROM riwayat_pembayaran WHERE nisn = '{nisn}' LIMIT 1"
            );
            string pembayaran = JsonConvert.SerializeObject(dtPembayaran);

            string queryBase =
                $"SELECT p.id_pembayaran, p.bulan_dibayar, p.tgl_bayar, t.nama_petugas " +
                $"FROM pembayaran p " +
                $"INNER JOIN petugas t ON p.id_petugas = t.id_petugas " +
                $"WHERE p.nisn = '{nisn}'";

            var bulan = new[] {
                "Juli","Agustus","September","Oktober","November","Desember",
                "Januari","Februari","Maret","April","Mei","Juni"
            };

            List<Dictionary<string, string>> hasil = new List<Dictionary<string, string>>();

            foreach (var b in bulan)
            {
                DataTable dt = db.Query(queryBase + $" AND bulan_dibayar = '{b}'");
                Dictionary<string, string> obj;

                if (dt.Rows.Count == 0)
                {
                    obj = new Dictionary<string, string>
                    {
                        ["pesan"] = "Belum Bayar",
                        ["id_pembayaran"] = "",
                        ["bulan_dibayar"] = b,
                        ["tgl_bayar"] = "",
                        ["nama_petugas"] = "",
                    };
                }
                else
                {
                    var row = dt.Rows[0];
                    obj = new Dictionary<string, string>
                    {
                        ["pesan"] = "",
                        ["id_pembayaran"] = row["id_pembayaran"].ToString(),
                        ["bulan_dibayar"] = b,
                        ["tgl_bayar"] = row["tgl_bayar"].ToString(),
                        ["nama_petugas"] = row["nama_petugas"].ToString(),
                    };
                }

                hasil.Add(obj);
            }

            return Ok(new
            {
                pembayaran,
                hasil
            });
        }

        [HttpGet("cetak/{id_pembayaran}")]
        public IActionResult CetakInvoice(string id_pembayaran)
        {
            DB db = new DB();

            DataTable dt = db.Query($"CALL cetak_invoice('{id_pembayaran}')");
            string invoice = JsonConvert.SerializeObject(dt);

            return Ok(new { invoice });
        }

        [HttpGet("bulan/{nisn}")]
        public IActionResult GetBulan(string nisn)
        {
            DB db = new DB();

            string queryBase = $"SELECT id_pembayaran, bulan_dibayar FROM pembayaran WHERE nisn = '{nisn}' AND deleted_at IS NULL";

            var bulan = new[] {
                "Juli","Agustus","September","Oktober","November","Desember",
                "Januari","Februari","Maret","April","Mei","Juni"
            };

            List<string> bulanList = new List<string>();

            foreach (var b in bulan)
            {
                DataTable dt = db.Query(queryBase + $" AND bulan_dibayar = '{b}'");

                if (dt.Rows.Count == 0)
                {
                    bulanList.Add(b);
                }
            }

            return Ok(new { bulan = bulanList });
        }
    }
}
